### ISSUE TITLE
crash in QRingBuffer::reserve

### ISSUE THREAD

**Issue by stefansaraev** [hyperion-project/hyperion.ng#133]
hi there,

today I tried runing a self compiled hyperion.ng on my wetek core and it fails to run
```
TB:~/.kodi/addons/service.hyperion/bin # gdb --args ./hyperiond /storage/.kodi/userdata/addon_data/service.hyperion/hyperion.config.jso
n 
...
Program received signal SIGSEGV, Segmentation fault.
QRingBuffer::reserve (this=0x0, bytes=512) at tools/qringbuffer.cpp:111
111	tools/qringbuffer.cpp: No such file or directory.
(gdb) thread apply all bt

Thread 1 (process 520):
#0  QRingBuffer::reserve (this=0x0, bytes=512) at tools/qringbuffer.cpp:111
#1  0x001cf2e6 in QSerialPortPrivate::readNotification() ()
#2  0x001cf8e8 in QSerialPortPrivate::waitForBytesWritten(int) ()
Backtrace stopped: previous frame identical to this frame (corrupt stack?)
```

I dont know if it's qt or hyperiond bug, but I decided to post this here.

```
[HYPERIOND LedDevice] <ERROR> Unable to open RS232 device ()
```
^^ I've got this once in a gdb session, just once. no matter how hard I try, I cant get it anymore

compiling against qtbase-5.7.0 + qtserialport-5.7.0, there is an extra (it should not interfere in no way) patch to make avahi optional. I can PR / share it if you want me to.

my compile opts: http://sprunge.us/LMXD
my config: http://sprunge.us/GRMa

device is a wetek core + dyi arduino nano clone + ws2812b leds (adalight). it works fine with https://github.com/hyperion-project/hyperion self compiled from master branch.

**Comment by redPanther** [comment]
here the same thing:
https://github.com/hyperion-project/hyperion.ng/issues/87

big thanx for that good bug report!

**Comment by stefansaraev** [comment]
ouch. sorry for the dupe. I searched and somehow missed #87. fyi delayAfterConnect was my first guess, I played a bit with different values but it did not help..

**Comment by redPanther** [comment]
it looks like calling waitforbyteswritten when it is not neccessary can lead to problems. Waiting is only necessary when write returns with "not all written yet".

here some idea:
http://stackoverflow.com/questions/35273827/serial-write-crashes-program-in-qt-c

I will modify code to something like:
```
    qint64 bytesWritten = serialPort.write(writeData);

    if (bytesWritten == -1) {
        standardOutput << QObject::tr("Failed to write the data to port %1, error: %2").arg(serialPortName).arg(serialPort.errorString()) << endl;
        return 1;
    } else if (bytesWritten != writeData.size()) {
        standardOutput << QObject::tr("Failed to write all the data to port %1, error: %2").arg(serialPortName).arg(serialPort.errorString()) << endl;
        return 1;
    } else if (!serialPort.waitForBytesWritten(5000)) {
        standardOutput << QObject::tr("Operation timed out or an error occurred for port %1, error: %2").arg(serialPortName).arg(serialPort.errorString()) << endl;
        return 1;
    }
```

**Comment by stefansaraev** [comment]
k http://sprunge.us/DZXR with this no crash and amlgrabber seems to work now. framegrabber (used for software decoded content in kodi) does not work, startup effect does not work. but thats another story :)

**Comment by brindosch** [comment]
It would be great if someone could extend the amlgrabber to work also with kodi. Or is Kodi a workaround? It would be hard with 2 grabbers at the same. As our target is one grabber for one system and not two :)
> startup effect does not work

checkout the current valid config at the config dir

**Comment by stefansaraev** [comment]
amlgrabber works "behind" kodi's back (takes frames fron /dev/amvideocap). because amcodec renders behind kodi's back. it is more like a v4l grabber, and it always worked for me. still works :)

previously, I used different priorities. 900 for bootsequence / framegrabber, and 800 for amlgrabber and it worked nicely. however, this is unrelated and can be looked at later

**Comment by brindosch** [comment]
No, i wanted to point to this:
https://github.com/hyperion-project/hyperion.ng/blob/master/config/hyperion.config.json.default
This should be your basic file for hyperion.ng and will be updated whenever we change something.

**Comment by stefansaraev** [comment]
@brindosch yep, I failed to update my old config (sorry for the extra noise). ./hyperion-remote -e worked. adding the new initialEffect section does the trick.

**Comment by redPanther** [comment]
I did some deeper investigation. waitForBytesWritten causes after sending some bytes (7k) a resource error:
"An I/O error occurred when a resource becomes unavailable, e.g. when the device is unexpectedly removed from the system."

I guess in your code waitForBytesWritten would never be executed and thatswyh it worked.

In above example result != size is an error, in your code would waitForBytesWritten be executed in that case.

I removed waitForBytesWritten, because it is not neccesary - the flush blocks until all data is written to port and do nearly the same.
